graphbolt/src/cuda/unique_and_compact_impl.cu

```diff
@@ -41,139 +44,254 @@ struct EqualityFunc {
+              gather, new_src_ids_loc[i].get(),
+              new_src_ids_loc[i].get() + src_ids[i].size(0),
+              new_ids[i].data_ptr<int64_t>(),
+              new_src_ids[i].data_ptr<index_t>());
+        }
         // Perform check before we gather for the dst indices.
-        if (dst_ids.size(0) > 0 && !static_cast<bool>(all_exist)) {
-          throw std::out_of_range("Some ids not found.");
-        }
-        auto new_dst_ids = torch::empty_like(dst_ids);
-        THRUST_CALL(
-            gather, new_dst_ids_loc.get(),
-            new_dst_ids_loc.get() + dst_ids.size(0),
-            new_ids.data_ptr<int64_t>(), new_dst_ids.data_ptr<scalar_t>());
-        return std::make_tuple(real_order, new_src_ids, new_dst_ids);
+        for (std::size_t i = 0; i < dst_ids.size(); i++) {
+          if (dst_ids[i].size(0) > 0) {
+            if (should_record) {
+              all_exist_event.synchronize();
+              should_record = false;
+            }
+            if (!static_cast<bool>(all_exist[i])) {
+              throw std::out_of_range("Some ids not found.");
+            }
+          }
+        }
+        std::vector<torch::Tensor> new_dst_ids;
+        for (std::size_t i = 0; i < dst_ids.size(); i++) {
+          new_dst_ids.emplace_back(torch::empty_like(dst_ids[i]));
+          THRUST_CALL(
+              gather, new_dst_ids_loc[i].get(),
+              new_dst_ids_loc[i].get() + dst_ids[i].size(0),
+              new_ids[i].data_ptr<int64_t>(),
+              new_dst_ids[i].data_ptr<index_t>());
+        }
+        std::vector<std::tuple<torch::Tensor, torch::Tensor, torch::Tensor>>
+            results;
+        for (std::size_t i = 0; i < src_ids.size(); i++) {
+          results.emplace_back(
+              std::move(real_order[i]), std::move(new_src_ids[i]),
+              std::move(new_dst_ids[i]));
+        }
+        return results;
       }));
 }
 
+std::vector<std::tuple<torch::Tensor, torch::Tensor, torch::Tensor>>
+UniqueAndCompactBatched(
+    const std::vector<torch::Tensor>& src_ids,
+    const std::vector<torch::Tensor>& dst_ids,
+    const std::vector<torch::Tensor>& unique_dst_ids, int num_bits) {
+  auto dev_id = cuda::GetCurrentStream().device_index();
+  static std::mutex mtx;
+  static std::unordered_map<decltype(dev_id), int> compute_capability_cache;
+  const auto compute_capability_major = [&] {
+    std::lock_guard lock(mtx);
+    auto it = compute_capability_cache.find(dev_id);
+    if (it != compute_capability_cache.end()) {
+      return it->second;
+    } else {
+      int major;
+      CUDA_DRIVER_CHECK(cuDeviceGetAttribute(
+          &major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, dev_id));
+      return compute_capability_cache[dev_id] = major;
+    }
+  }();
+  if (compute_capability_major >= 7) {
+    return UniqueAndCompactBatchedMap(src_ids, dst_ids, unique_dst_ids);
+  }
+  return UniqueAndCompactBatchedSort(
+      src_ids, dst_ids, unique_dst_ids, num_bits);
```

Comment:
Add comment to explain the difference between Sort & Map?

Reply:
Added.